### PR TITLE
Add dockerfile and Cargo.lock

### LINKED ISSUE #8: Consider committing Cargo.lock for reproducibility
The Cargo book suggest to commit `Cargo.lock` for binaries. See [this](https://doc.rust-lang.org/cargo/faq.html#why-do-binaries-have-cargolock-in-version-control-but-not-libraries).

### PR DESCRIPTION
**Pull request by marcoieni**
Closes https://github.com/d-e-s-o/cargo-http-registry/issues/8

### CODE REVIEW

**Comment by d-e-s-o** [comment]
Thanks! Will take a closer look later today or tomorrow.

File: .github/workflows/docker.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      - name: Cache Docker layers
+        uses: actions/cache@v3
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
+
+      - name: Login to GHCR
+        if: github.event_name != 'pull_request'
```
Comment (d-e-s-o) [inline .github/workflows/docker.yml]: Isn't this condition always true? The trigger is a push, right?

File: .github/workflows/docker.yml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            type=sha
+            type=raw,value=latest
+
+      - name: Cache Docker layers
+        uses: actions/cache@v3
+        with:
+          path: /tmp/.buildx-cache
+          key: ${{ runner.os }}-buildx-${{ github.sha }}
+          restore-keys: |
+            ${{ runner.os }}-buildx-
```
Comment (d-e-s-o) [inline .github/workflows/docker.yml]: For my understanding, is this caching actually necessary? And does it help much when used? I don't expect that we have too frequent commits, to the point that cached artifacts probably have a high likelihood of expiring in between. Given all the workarounds we need below (which chances are will never get cleaned up), perhaps we can just simplify things by removing caching? From my perspective, I don't particularly care much how long GitHub Actions takes once a push to the main branch has happened -- it's not blocking a pull request. So simpler seems better. What do you think?

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,23 @@
+FROM lukemathwalker/cargo-chef:0.1.59-rust-bullseye AS chef
+WORKDIR /app
+
+FROM chef as planner
+COPY . .
```
Comment (d-e-s-o) [inline Dockerfile:5]: Out of curiosity, what is this line for?

File: .github/workflows/docker.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        uses: docker/login-action@v2
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Build and push
+        uses: docker/build-push-action@v4
+        with:
+          platforms: linux/amd64,linux/arm64
```
Comment (d-e-s-o) [inline .github/workflows/docker.yml:47]: We specify two platforms here, which makes sense. How did the `unknown/unknown` combo, which I see on your fork, come into being, though?

File: .github/workflows/docker.yml [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+          push: ${{ github.event_name != 'pull_request' }}
```
Comment (d-e-s-o) [inline .github/workflows/docker.yml]: Similar comment to above.

File: .github/workflows/docker.yml [hunk 6]
```diff
@@ -0,0 +1,68 @@
+name: Docker
+
+on:
+  push:
```
Comment (d-e-s-o) [inline .github/workflows/docker.yml:5]: Shall we also add `workflow_dispatch:` as a trigger to be able to manually redo? Or is the registry read-only for the same SHA or something?

**Review by d-e-s-o — COMMENTED** [review]
Overall seems fine to me, thanks! Left few comments/questions.

File: .github/workflows/docker.yml (hunk 1 above)
Comment (marcoieni) [reply .github/workflows/docker.yml]: removed in [093ad4c](https://github.com/d-e-s-o/cargo-http-registry/pull/9/commits/093ad4c892c12093e3c251dbed3901f41c62e7b2)

File: .github/workflows/docker.yml (hunk 5 above)
Comment (marcoieni) [reply .github/workflows/docker.yml]: done in [093ad4c](https://github.com/d-e-s-o/cargo-http-registry/pull/9/commits/093ad4c892c12093e3c251dbed3901f41c62e7b2)

File: .github/workflows/docker.yml (hunk 6 above)
Comment (marcoieni) [reply .github/workflows/docker.yml:5]: I don't think it's read-only, I added the `workflow_dispatch` in [c1b15e9](https://github.com/d-e-s-o/cargo-http-registry/pull/9/commits/c1b15e9ebfb001bdf84c5fdaf9302bfb25e5c5c6)

I will test it

File: .github/workflows/docker.yml (hunk 4 above)
Comment (marcoieni) [reply .github/workflows/docker.yml:47]: I found [this](https://github.com/orgs/community/discussions/45969) discussion.

let's see if https://github.com/MarcoIeni/cargo-http-registry/commit/b55bec101844db61cacf2c61265b90df4a590e36 fixes it.

File: Dockerfile (hunk 3 above)
Comment (marcoieni) [reply Dockerfile:5]: It copies the current directory (i.e. the application itself) in the container.

I followed [this](https://github.com/LukeMathWalker/cargo-chef#how-to-use) guide.

File: .github/workflows/docker.yml (hunk 2 above)
Comment (marcoieni) [reply .github/workflows/docker.yml]: makes sense! Done in [b2f5d86](https://github.com/d-e-s-o/cargo-http-registry/pull/9/commits/b2f5d865f0c428ad8f0351aef15ebe17f5eba673)

File: .github/workflows/docker.yml (hunk 4 above)
Comment (marcoieni) [reply .github/workflows/docker.yml:47]: it worked!
image

File: .github/workflows/docker.yml (hunk 6 above)
Comment (marcoieni) [reply .github/workflows/docker.yml:5]: The manual run worked: https://github.com/MarcoIeni/cargo-http-registry/actions/runs/5053965210

File: .github/workflows/docker.yml (hunk 4 above)
Comment (d-e-s-o) [reply .github/workflows/docker.yml:47]: Very cool, thanks for getting to the bottom of it and fixing it! :+1:

**Review by d-e-s-o — APPROVED** [review]
Great work! Will give it a spin myself and then merge. Thanks again!

**Comment by d-e-s-o** [comment]
I cleaned up the history a bit, added CHANGELOG entries, and merged. Will set up dependabot soon. Thanks!

**Pull request closed without merging**